### PR TITLE
LinuxBrew with PIE: Permissions get wrongly 'cleaned'

### PR DESCRIPTION
**Pull request by thriqon**
**tl;dr**: `alpine$ brew install hello && brew test hello` fails with insufficient permissions due to PIE.

Alpine Linux (and possibly others) produce [Position Independent Executables](https://en.wikipedia.org/wiki/Position-independent_code), which take the form of shared libraries.
However, from an ELF and therefore from a `path.mach_o_executable?` point of view those are not really `executable`s, which leads to them wrongfully getting `chmod 0444`ed.

## Steps to reproduce

    cat > Dockerfile <<EOF
    FROM alpine
    RUN apk update &&\
      apk add git make clang ruby ruby-irb ncurses tar binutils build-base &&\
      git clone https://github.com/Homebrew/linuxbrew.git ~/brew
    RUN /root/brew/bin/brew install hello
    RUN /root/brew/bin/brew test hello
    EOF
    docker build .

## Suggestion to fix

There is no harm in having libraries executable. In fact, they usually _are_ marked executable anyways. I suggest expanding the check in [Library/Homebrew/cleaner.rb:90](https://github.com/Homebrew/linuxbrew/blob/master/Library/Homebrew/cleaner.rb#L90)

```ruby
perms = if path.mach_o_executable? || path.text_executable?
```

to

```ruby
perms = if path.mach_o_executable? || path.dylib? || path.text_executable?
```

EDIT: I'm unable to run the tests on my machine... It seems to expect to be run on a Mac?

### CODE REVIEW

**Comment by thriqon** [comment]
Shall I request this on Homebrew/homebrew instead?

**Comment by DoomHammer** [comment]
I believe this belongs here and is of not much value to Homebrew. @sjackman is travelling right now so might be pretty slow reviewing PRs.

**Comment by sjackman** [comment]
That's unusual to use PIE for executables. Why does Alpine do that?

I don't see any harm in applying this patch, but it'll have to wait until I'm back from vacation in the new year. Feel free to bump this thread then.

**Comment by thriqon** [comment]
PIE is used for Address Space Layout Randomization. It's used to protect against attacks based on knowledge of the memory layout (buffer overflows etc). AFAIK the position of `main` is shifted around as well, which makes it impossible to use 'normal' executables.

I whish you a nice travel! I'll bump it again then...

**Comment by sjackman** [comment]
Ah, yes. I actually thought that was more common even with Redhat and Debian. Perhaps only for shared libraries.

**Comment by sjackman** [comment]
Merged, making sure to keep the behaviour unchanged on Mac OS: dylibs should be executable on Linux but not on Mac. Thanks for your contribution, Jonas!
